### PR TITLE
[IMPROVE] App session timeout handler 

### PR DESCRIPTION
**Pull request by AlexanderKanakis**
Added App session timeout handler and created app setting for switching between app and widget session timeout handlers

### CODE REVIEW

File: helperFunctions/TimeoutHelper.ts [hunk 1]
```diff
@@ -42,38 +42,50 @@ export const handleTimeout = async (app: IApp, message: IMessage, read: IRead, h
...
+					idleTimeoutAction: 'stop',
+					idleTimeoutWarningTime: warningTime,
+					idleTimeoutTimeoutTime: timeoutTime,
+					idleTimeoutMessage: timeoutWarningMessage,
+				});
+				(await msgExtender).addCustomField('sneakPeekEnabled', sneakPeekEnabled);
+				modify.getExtender().finish(await msgExtender);
+			} else {
+				await persistence.createWithAssociation({ idleSessionScheduleStarted: false }, assoc);
+				modify.getScheduler().cancelJob('idle-session-timeout');
```
Comment (Shailesh351) [inline helperFunctions/TimeoutHelper.ts]: `cancelJob` will cancel all the jobs (for all the chats). Here it should only cancel the job for a particular room

File: SalesforceLiveAgentApp.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ import { AppSettings } from './config/AppSettings';
 import { AvailabilityEndpoint } from './endpoints/AvailabilityEndpoint';
 import { HandoverEndpoint } from './endpoints/HandoverEndpoint';
 import { DialogflowAgentAssignedClass } from './lib/DialogflowAgentAssignedHandler';
+import { IdleSessionTimeoutProcessor } from './lib/IdleSessionTimeoutProcessor';
```
Comment (Shailesh351) [inline SalesforceLiveAgentApp.ts:22]: I think you forgot to add `IdleSessionTimeoutProcessor` file in the `lib` folder 😜

**Comment by ear-dev** [comment]
@Shailesh351 @AlexanderKanakis where are we with this PR? 

This will replace the session timeout on the widget?  So that the widget timeout value is still valid during a DF session, but it then becomes disabled during a SF session?  Theoretically this should allow a user to sit in a queue for any configured length of time?

**Comment by ear-dev** [comment]
@Shailesh351 @AlexanderKanakis is this one ready to be tested?

**Comment by AlexanderKanakis** [comment]
@ear-dev yes, from my part.

**Comment by Shailesh351** [comment]
@ear-dev I didn’t get a chance to test it, but from the code looks like it won't show the `warning timer` and `Chat ended due to timeout` Popup.

**Comment by AlexanderKanakis** [comment]
@Shailesh351 @ear-dev These messages cannot appear on non-widget clients so I just included a functional timer. I can make some slight changes on Apps.Dialogflow and the Rocket.chat.Livechat to accomodate for it.

**Comment by ear-dev** [comment]
@Shailesh351 @AlexanderKanakis @chadgoss I believe that after our meeting this morning we decided that we do not actually want the warning timer or the popup.  Do you think that's correct? 

However, it might be nice to have a configurable message we send in the chat room when the timer runs out......  something akin to "Custom message when room is automatically closed by visitor inactivity"....  what do you think?

In the meantime, I'll deploy this to our preprod and test.  Thanks @AlexanderKanakis !

**Comment by ear-dev** [comment]
@AlexanderKanakis I've deployed this version of the SF app, and I can see the switch in the UI, but where is the timer value set in this case?

image

**Comment by AlexanderKanakis** [comment]
We receive the value from the chat button we receive from salesforce. The button contains a session warning and a session timeout setting. 
image

**Comment by ear-dev** [comment]
@AlexanderKanakis @Shailesh351 @chadgoss I think this one looks good to go.....  I'm just waiting for one more confirmation from Chad, then I'll merge.  

We'll create a new story for "Room close event triggered on timeouts - VA (omnichannel SF Dept), Customer Timeout Value"

**Pull request merged**